contracts/wallet/AdminWallet.sol

```diff
@@ -0,0 +1,147 @@
+
+pragma solidity ^0.5.4;
+
+import "openzeppelin-solidity/contracts/access/Roles.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "../token/GoodDollar.sol";
+import "../identity/Identity.sol";
+
+/* @title Admin wallet contract allowing whitelisting and topping up of
+ * addresses
+ */
+contract AdminWallet is Ownable {
+    using Roles for Roles.Role;
+    using SafeMath for uint256;
+
+    Roles.Role private admins;
+
+    GoodDollar token;
+    Identity identity;
+
+    uint256 public toppingAmount;
+    
+    uint public toppingTimes;
+    uint public lastCalc;
+
+    mapping(uint => mapping(address => uint)) toppings;
+
+    event AdminsAdded(address[] indexed admins);
+    event AdminsRemoved(address[] indexed admins);
+    event WalletTopped(address indexed user);
+
+    constructor (
+        address[] memory _admins,
+        uint256 _toppingAmount,
+        uint _toppingTimes,
+        Identity _identity
+    )
+        public 
+    {
+        identity = _identity;
+
+        toppingAmount = _toppingAmount;
+        toppingTimes = _toppingTimes;
+
+        addAdmins(_admins);
+    }   
+
+    /* @dev Modifier that checks if caller is admin of wallet
+     */
+    modifier onlyAdmin() {
+        require(isAdmin(msg.sender), "Caller is not admin");
+        _;
+    }
+
+    modifier reimburseGas() {
+        _;
+        require(msg.sender.balance <= toppingAmount.div(4), "Admin balance too high");
```

Comment:
This is going to generate exception and prevent normal operation if wallet doesn't require topping